<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function normalizes the genomic coordinates of a set of genes within a
cluster, ensuring that the genes are evenly spaced along the entire range of
the cluster. The function allows for the option to preserve the original gene
lengths and to introduce customized gaps between the genes."><title>Normalize Gene Clusters in a Genomic Chart — GC_normalize • geneviewer</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Normalize Gene Clusters in a Genomic Chart — GC_normalize"><meta property="og:description" content="This function normalizes the genomic coordinates of a set of genes within a
cluster, ensuring that the genes are evenly spaced along the entire range of
the cluster. The function allows for the option to preserve the original gene
lengths and to introduce customized gaps between the genes."><meta property="og:image" content="https://nvelden.github.io/geneviewer/logo.png"><meta name="google-site-verification" content="yTGAUIDwbj1rCz39OTZSzq5wZAPsbzfL9O_Gc7u6w14"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- dependencies from examples --><script src="libs/htmlwidgets-1.6.4/htmlwidgets.js"></script><script src="libs/geneviewerwidget-0.1.8/geneviewerwidget.js"></script><script src="libs/Themes-0.1.8/./lib/geneviewer-0.1.8/Themes.js"></script><script src="libs/geneviewer-0.1.8/./lib/geneviewer-0.1.8/geneviewer.js"></script><link href="libs/D3-7.8.5/./lib/styles.css" rel="stylesheet"><script src="libs/D3-7.8.5/./lib/D3-7.8.5/d3.min.js"></script><script src="libs/html2canvas-1.4.1/./lib/html2canvas-1.4.1/html2canvas.min.js"></script><script src="libs/FileSaver-2.0.4-2.0.4/./lib/FileSaver-2.0.4/FileSaver.min.js"></script><link href="libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">geneviewer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/geneviewer.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/BLASTP.html">Cluster comparison using BlastP</a>
    <a class="dropdown-item" href="../articles/Examples.html">Example charts</a>
    <a class="dropdown-item" href="../articles/LoadFastaFiles.html">Loading Gene Clusters From FASTA Files</a>
    <a class="dropdown-item" href="../articles/LoadGenBankFiles.html">Loading Gene Clusters From Genbank Files</a>
    <a class="dropdown-item" href="../articles/LoadGFF.html">Loading Gene Clusters From Generic Feature Format (GFF) Files</a>
    <a class="dropdown-item" href="../articles/MUMMER.html">Genome alignment using MUMmer</a>
    <a class="dropdown-item" href="../articles/Transcripts.html">Visualizing gene Transcripts</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nvelden/geneviewer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Normalize Gene Clusters in a Genomic Chart</h1>
      <small class="dont-index">Source: <a href="https://github.com/nvelden/geneviewer/blob/HEAD/R/geneviewer.R" class="external-link"><code>R/geneviewer.R</code></a></small>
      <div class="d-none name"><code>GC_normalize.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function normalizes the genomic coordinates of a set of genes within a
cluster, ensuring that the genes are evenly spaced along the entire range of
the cluster. The function allows for the option to preserve the original gene
lengths and to introduce customized gaps between the genes.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GC_normalize</span><span class="op">(</span></span>
<span>  <span class="va">GC_chart</span>,</span>
<span>  group <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cluster <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  preserve_gene_length <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  gap <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  custom_gaps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>GC_chart</dt>
<dd><p>A chart object containing genomic data and clustering
information. The chart object must include a `series` component with data
for each cluster.</p></dd>


<dt>group</dt>
<dd><p>Column name containing gene names to apply custom gaps to.</p></dd>


<dt>cluster</dt>
<dd><p>A vector of cluster names to be normalized. If `NULL`, all
clusters in the `GC_chart` will be normalized.</p></dd>


<dt>normalize</dt>
<dd><p>A logical value indicating whether to normalize the gene
positions within each cluster. Default is `TRUE`.</p></dd>


<dt>preserve_gene_length</dt>
<dd><p>A logical vector indicating whether to preserve
the original gene lengths for each cluster. If a single value is provided,
it is recycled for all clusters.</p></dd>


<dt>gap</dt>
<dd><p>A numeric vector specifying the proportion of the total length to
be used as the gap between genes in each cluster. The value must be between
0 and 1. If a single value is provided, it is recycled for all clusters. If
`NULL`, the gaps are calculated based on the actual spacing between the
genes in the original data.</p></dd>


<dt>custom_gaps</dt>
<dd><p>A named list where each name corresponds to a gene name and
each value is a numeric factor by which to adjust the gap after that gene.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The modified `GC_chart` object with normalized gene coordinates for
  the specified clusters.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">genes_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">90</span>, <span class="fl">130</span>, <span class="fl">10</span>, <span class="fl">90</span>, <span class="fl">130</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">40</span>, <span class="fl">120</span>, <span class="fl">160</span>, <span class="fl">40</span>, <span class="fl">120</span>, <span class="fl">160</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Gene 1'</span>, <span class="st">'Gene 2'</span>, <span class="st">'Gene 3'</span>,<span class="st">'Gene 1'</span>, <span class="st">'Gene 2'</span>, <span class="st">'Gene 3'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="st">'B'</span>, <span class="st">'C'</span>, <span class="st">'A'</span>, <span class="st">'B'</span>, <span class="st">'C'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  cluster <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="GC_chart.html">GC_chart</a></span><span class="op">(</span><span class="va">genes_data</span>, group <span class="op">=</span> <span class="st">"group"</span>, cluster <span class="op">=</span> <span class="st">"cluster"</span>, height <span class="op">=</span> <span class="st">"150px"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">GC_normalize</span><span class="op">(</span></span></span>
<span class="r-in"><span>    cluster <span class="op">=</span> <span class="fl">2</span>,</span></span>
<span class="r-in"><span>    preserve_gene_length <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    gap <span class="op">=</span> <span class="cn">NULL</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<div id="htmlwidget-da403bf8187c892ade73" style="width:100%;height:150px;" class="geneviewer html-widget"></div>
<script type="application/json" data-for="htmlwidget-da403bf8187c892ade73">{"x":{"params":{"start":"start","end":"end","cluster":"cluster","group":"group","width":"100%","height":"150px","save_button":true},"data":{"rowID":[1,2,3,4,5,6],"start":[10,90,130,10,60,110],"end":[40,120,160,40,90,140],"name":["Gene 1","Gene 2","Gene 3","Gene 1","Gene 2","Gene 3"],"group":["A","B","C","A","B","C"],"cluster":["1","1","1","2","2","2"],"strand":["forward","forward","forward","forward","forward","forward"],"original_start":[null,null,null,10,90,130],"original_end":[null,null,null,40,120,160]},"group":"group","cluster":"cluster","graphContainer":{"direction":"column"},"style":[],"title":{"style":{"width":"100%"}},"legend":{"group":"group","show":true,"position":"bottom","width":"100%","style":{"width":"100%","backgroundColor":null}},"saveButton":true,"series":{"1":{"clusterName":"1","data":{"start":[10,90,130],"end":[40,120,160],"name":["Gene 1","Gene 2","Gene 3"],"group":["A","B","C"],"cluster":["1","1","1"],"rowID":[1,2,3],"strand":["forward","forward","forward"]},"cluster":[],"container":{"height":75,"width":"100%","style":{"width":"100%","backgroundColor":null}},"genes":{"group":"group","show":true},"sequence":{"show":true},"scale":{"xMin":10,"xMax":160},"transcript":{"group":"group","show":false},"labels":{"group":"group","show":true},"coordinates":{"show":false},"scaleBar":[],"footer":[],"clusterLabel":[],"clusterTitle":[],"annotations":[],"trackMouse":{"show":false},"tooltip":{"show":true,"formatter":"<b>Start:<\/b> {start}<br><b>End:<\/b> {end}"}},"2":{"clusterName":"2","data":{"start":[10,60,110],"end":[40,90,140],"name":["Gene 1","Gene 2","Gene 3"],"group":["A","B","C"],"cluster":["2","2","2"],"rowID":[4,5,6],"strand":["forward","forward","forward"],"original_start":[10,90,130],"original_end":[40,120,160]},"cluster":[],"container":{"height":75,"width":"100%","style":{"width":"100%","backgroundColor":null}},"genes":{"group":"group","show":true},"sequence":{"show":true},"scale":{"xMin":10,"xMax":140},"transcript":{"group":"group","show":false},"labels":{"group":"group","show":true},"coordinates":{"show":false},"scaleBar":[],"footer":[],"clusterLabel":[],"clusterTitle":[],"annotations":[],"trackMouse":{"show":false},"tooltip":{"show":true,"formatter":"<b>Start:<\/b> {start}<br><b>End:<\/b> {end}"}}}},"evals":[],"jsHooks":[]}</script><span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Niels van der Velden.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

